compiler/rustc_builtin_macros/src/proc_macro_harness.rs

```diff
@@ -202,26 +214,94 @@ impl<'a> Visitor<'a> for CollectProcMacros<'a> {
-            return;
+        let mut extra_attrs = Vec::new();
+
+        match kind {
+            ProcMacroAttr::Derive | ProcMacroAttr::Attribute | ProcMacroAttr::Bang => {
+                if !matches!(item_inner.kind, ast::ItemKind::Fn(..)) {
+                    self.dcx
+                        .create_err(errors::AttributeOnlyBeUsedOnBareFunctions {
+                            span: attr.span,
+                            path: &kind.to_string(),
+                        })
+                        .emit();
+                    return;
+                }
+            }
+            ProcMacroAttr::Lint => {
+                let ast::ItemKind::Static(static_item) = &mut item_inner.kind else {
+                    self.dcx
+                        .create_err(errors::AttributeOnlyBeUsedOnStatics {
+                            span: attr.span,
+                            path: &kind.to_string(),
+                        })
+                        .emit();
+                    return;
+                };
+                if let Some(expr) = &static_item.expr {
+                    self.dcx
+                        .create_err(errors::LintIdIsFilledInByAttribute {
+                            span: static_item.ty.span.between(expr.span).to(expr.span),
+                        })
+                        .emit();
+                } else {
+                    // Expand `pub static lintname: LintId;` into:
+                    //
+                    //     #[allow(non_upper_case_globals)]
+                    //     pub static lintname: LintId = ::proc_macro::LintId::new("lintname");
+                    let expn_id = self.cx.resolver.expansion_for_ast_pass(
+                        attr.span,
+                        AstPass::ProcMacroHarness,
+                        // Edition >2015 to be able to use ::proc_macro without generating an `extern crate`.
+                        Some(rustc_span::edition::Edition::Edition2021),
+                        &[sym::proc_macro_internals],
+                        None,
+                    );
+                    let span = attr.span.with_def_site_ctxt(expn_id.to_expn_id());
+                    extra_attrs.push(attr::mk_attr_nested_word(
+                        &self.cx.psess().attr_id_generator,
+                        ast::AttrStyle::Outer,
+                        ast::Safety::Default,
+                        sym::allow,
+                        sym::non_upper_case_globals,
+                        span,
+                    ));
+                    let lintid_new = self.cx.expr_call(
+                        span,
+                        self.cx.expr_path(self.cx.path(
+                            span,
+                            vec![
+                                Ident::new(kw::PathRoot, span),
+                                Ident::new(sym::proc_macro, span),
+                                Ident::new(sym::LintId, span),
+                                Ident::new(sym::new, span),
```

Comment:
It should be better to make `LintId::new` a lang item and add the static initializer in AST lowering using `make_lang_item_path`.
This will allow to
- avoid a mutable AST visitor, their use should be minimized because they stand in the way of plans like keeping AST on arena, etc.
- avoid the edition hack above, lang item collection will be used to "resolve" the `LintId::new` path instead

Same for adding the `#[allow(non_upper_case_globals)]` attribute, it's better to adjust the lint directly.